### DISCUSSION TITLE
[ANSWERED] discord.app_commands.errors.MissingApplicationID Error

### DISCUSSION THREAD

**Discussion by Joshisaurio** [Rapptz/discord.py#9517] [Q&A] [answered]
## **Explanation**
I had made a bot just to have fun with my friends, and at first it would just detect whenever someone posted a message and replied, and it worked perfectly, but I wanted to add slash commands so of course I researched how to do it. And I'm pretty sure my code was right at that point, but I kept getting the discord.app_commands.errors.MissingApplicationID error whenever I tried to run the code.
Now, here is my code but it probably seems completely broken now since I basically tried everything and kind of broke the whole thing. Just want to note that whenever I remove everything that I added for the commands, it works, but as soon as I add it I get the error, but the bot keeps running anyway (I'm also using Keep_alive on replit but I removed it to save space). I also tried doing the same code for the commands in a separate bot and it worked for some reason. I even checked the intents and I don't think I was missing any, the only one I needed was the message content intent (which I needed for the bot to read the messages and reply) and it was checked already, and I'm pretty sure that the bot doesn't even need it. I hope someone can help!
Edit: I also tried to put the application id in the client constructor but it didn't work. It just kept generating the same thing or an error where it was missing intents. I don't even know what happened anymore or if it was the same error 

## **The code**
```
import discord
from discord import app_commands
from discord.ext import commands
import os

intents = discord.Intents.default()
intents.message_content = True

bot = commands.Bot(command_prefix="!", intents = intents)
client = discord.Client(intents=intents)

@client.event
async def on_ready():
  #print('Logged in as {0.user}'.format(client))
  print('Bot is ready!')
  #try:
  synced = await bot.tree.sync()
  print(f"Synced {len(synced)} command(s)")
  #except Exception as e:
    #print(e)

@client.event
async def on_message(message):
  if message.author == client.user:
      return

  messagecontent = str(message.content).lower()
  
  if "ola" in messagecontent:
        print("ALGUIEN DIJO OLA >:(")
        await message.channel.send('se dice hola :v')
        await message.channel.send('<:picadie:1110627479950987294>')

@bot.tree.command(name="test")
async def test(interaction: discord.Interaction):
  await interaction.response.send_message(f"Hola {interaction.user.mention}, recuerda no decir Ola porque me enojo", ephemeral = True)

try:
  client.run(os.getenv("TOKEN"))
  bot.run(os.getenv("TOKEN"))
except discord.HTTPException as e:
  if e.status == 429:
      print("The Discord servers denied the connection for making too many requests")
      print("Get help from https://stackoverflow.com/questions/66724687/in-discord-py-how-to-solve-the-error-for-toomanyrequests")
  else:
      raise e
```

## **The full traceback:**
```
Traceback (most recent call last):
  File "/home/runner/Yet-Another-Random-Bot/venv/lib/python3.8/site-packages/discord/client.py", line 441, in _run_event
    await coro(*args, **kwargs)
  File "main.py", line 23, in on_ready
    synced = await bot.tree.sync()
  File "/home/runner/Yet-Another-Random-Bot/venv/lib/python3.8/site-packages/discord/app_commands/tree.py", line 1055, in sync
    raise MissingApplicationID
discord.app_commands.errors.MissingApplicationID: Client does not have an application_id set. Either the function was called before on_ready was called or application_id was not passed to the Client constructor.
```

**Comment by LostLuma** [comment, marked as answer]
Hello, your issue stems from you defining both a `Client` and `Bot` instance and mixing them up afterwards. The `Bot` class is an extension of the `Client` class, and if you'd like to switch from one to the other to make use of additional functionality in Bot you should *replace* the Client with a Bot, rather than creating it in addition to the other.

The error specifically occurs because the internal `application_id` field is filled when the library connects to Discord, and you're syncing the bot's `CommandTree` within the ready event for the `client`, rather than the one for your `bot`. If you only have one of the two as described above your issue should resolve itself 🙂 Do note that the only proper way of fixing this is to only have one of the two, running multiple applications within a single program is not recommended and will likely lead you to even more confusing issues.

**Reply by Joshisaurio** [reply]
Thanks! I'll try it when I get back to my pc

**Reply by Joshisaurio** [reply]
This worked! Tysm
